include/LightGBM/cuda/cuda_algorithms.hpp

```diff
@@ -0,0 +1,392 @@
+__device__ __forceinline__ T ShuffleReduceSum(T value, T* shared_mem_buffer, const size_t len) {
+  const uint32_t warpLane = threadIdx.x % warpSize;
+  const uint32_t warpID = threadIdx.x / warpSize;
+  const data_size_t warp_len = min(static_cast<data_size_t>(warpSize), static_cast<data_size_t>(len) - static_cast<data_size_t>(warpID * warpSize));
+  value = ShuffleReduceSumWarp<T>(value, warp_len);
+  if (warpLane == 0) {
+    shared_mem_buffer[warpID] = value;
+  }
+  __syncthreads();
+  const data_size_t num_warp = static_cast<data_size_t>((len + warpSize - 1) / warpSize);
+  if (warpID == 0) {
+    value = (warpLane < num_warp ? shared_mem_buffer[warpLane] : 0);
+    value = ShuffleReduceSumWarp<T>(value, num_warp);
+  }
+  return value;
+}
+
+template <typename T>
+__device__ __forceinline__ T ShuffleReduceMaxWarp(T value, const data_size_t len) {
+  if (len > 0) {
+    // TODO(shiyu1994): check how mask works
+    const uint32_t mask = 0xffffffff;
+    for (int offset = warpSize / 2; offset > 0; offset >>= 1) {
+      value = max(value, __shfl_down_sync(mask, value, offset));
+    }
+  }
+  return value;
+}
+
+// reduce values from an 1-dimensional block (block size must be no greather than 1024)
+template <typename T>
+__device__ __forceinline__ T ShuffleReduceMax(T value, T* shared_mem_buffer, const size_t len) {
+  const uint32_t warpLane = threadIdx.x % warpSize;
+  const uint32_t warpID = threadIdx.x / warpSize;
+  const data_size_t warp_len = min(static_cast<data_size_t>(warpSize), static_cast<data_size_t>(len) - static_cast<data_size_t>(warpID * warpSize));
+  value = ShuffleReduceMaxWarp<T>(value, warp_len);
+  if (warpLane == 0) {
+    shared_mem_buffer[warpID] = value;
+  }
+  __syncthreads();
+  const data_size_t num_warp = static_cast<data_size_t>((len + warpSize - 1) / warpSize);
+  if (warpID == 0) {
+    value = (warpLane < num_warp ? shared_mem_buffer[warpLane] : 0);
+    value = ShuffleReduceMaxWarp<T>(value, num_warp);
+  }
+  return value;
+}
+
+// calculate prefix sum values within an 1-dimensional block in global memory, exclusively
+template <typename T>
+__device__ __forceinline__ void GlobalMemoryPrefixSum(T* array, const size_t len) {
+  const size_t num_values_per_thread = (len + blockDim.x - 1) / blockDim.x;
+  const size_t start = threadIdx.x * num_values_per_thread;
+  const size_t end = min(start + num_values_per_thread, len);
+  T thread_sum = 0;
+  for (size_t index = start; index < end; ++index) {
+    thread_sum += array[index];
+  }
+  __shared__ T shared_mem[32];
+  const T thread_base = ShufflePrefixSumExclusive<T>(thread_sum, shared_mem);
+  if (start < end) {
+    array[start] += thread_base;
+  }
+  for (size_t index = start + 1; index < end; ++index) {
+    array[index] += array[index - 1];
+  }
+}
+
+template <typename VAL_T, typename INDEX_T, bool ASCENDING>
+__device__ __forceinline__ void BitonicArgSort_1024(const VAL_T* scores, INDEX_T* indices, const INDEX_T num_items) {
```

Comment:
1024 ?

Reply:
1024 means that this version of BitonicArgSort only handles up to 1024 elements. Since this kernels is called only when the maximum number of bins in a features is no more than 256, BitonicArgSort_1024 suffices.